src/platform_impl/web/web_sys/resize_scaling.rs

```diff
@@ -0,0 +1,288 @@
+                if let Some(rc_self) = weak_self.upgrade() {
+                    let mut this = rc_self.borrow_mut();
+
+                    if this.resize_enabled.get() {
+                        let size = Self::process_entry(&this.window, &this.canvas, entries);
+                        (this.resize_handler)(size)
+                    }
+                }
+            });
+            let observer = Self::create_observer(&canvas, observer_closure.as_ref());
+
+            RefCell::new(Self {
+                window,
+                canvas,
+                mql,
+                observer,
+                _observer_closure: observer_closure,
+                scale_handler: Box::new(scale_handler),
+                resize_handler: Box::new(resize_handler),
+                resize_enabled: Cell::new(true),
+            })
+        })
+    }
+
+    fn create_mql<F>(window: &Window, closure: F) -> MediaQueryListHandle
+    where
+        F: 'static + FnMut(&MediaQueryList),
+    {
+        let current_scale = super::scale_factor(window);
+        // TODO: Remove `-webkit-device-pixel-ratio`. Requires Safari v16.
+        let media_query = format!(
+            "(resolution: {current_scale}dppx),
+             (-webkit-device-pixel-ratio: {current_scale})",
+        );
+        let mql = MediaQueryListHandle::new(window, &media_query, closure);
+        assert!(
+            mql.mql().matches(),
+            "created media query doesn't match, {current_scale} != {}",
+            super::scale_factor(window)
+        );
+        mql
+    }
+
+    fn create_observer(canvas: &HtmlCanvasElement, closure: &JsValue) -> ResizeObserver {
+        let observer = ResizeObserver::new(closure.as_ref().unchecked_ref())
+            .expect("Failed to create `ResizeObserver`");
+
+        // Safari doesn't support `devicePixelContentBoxSize`
+        if has_device_pixel_support() {
+            observer.observe_with_options(
+                canvas,
+                ResizeObserverOptions::new().box_(ResizeObserverBoxOptions::DevicePixelContentBox),
+            );
+        } else {
+            observer.observe(canvas);
+        }
+
+        observer
+    }
+
+    fn create_oneshot_observer(
+        &self,
+        weak_self: Weak<RefCell<Self>>,
+        handler: impl 'static + FnOnce(&mut Self, PhysicalSize<u32>),
+    ) {
+        self.resize_enabled.set(false);
+
+        let observer_closure =
+            Closure::once_into_js(move |entries: Array, observer: ResizeObserver| {
+                if let Some(rc_self) = weak_self.upgrade() {
+                    let mut this = rc_self.borrow_mut();
+                    let size = Self::process_entry(&this.window, &this.canvas, entries);
+                    this.resize_enabled.set(true);
+
+                    handler(&mut this, size)
+                }
+
+                observer.disconnect();
+            });
+        Self::create_observer(&self.canvas, &observer_closure);
+    }
```

Comment:
Hah, I actually had the same idea overnight and was about to suggest it, great minds think alike I guess. 👍 

One thing to be wary of though is that `ResizeObserver` isn't 100% guaranteed to call the callback with an initial resize - the way it's spec'd is that it's initialised with a last reported size of `(0, 0)`, and so if the canvas has a size of `(0, 0)` the callback won't fire.

It turns out Safari is the only browser that actually follows the spec here, though - Firefox and Chrome will always fire an initial resize. Specifically, Safari seems to not fire an initial resize for anything below about 0.018px.

You could probably also avoid having to create a temporary `ResizeObserver` here by calling `unobserve` then `observe` again on the canvas with the existing `ResizeObserver`, which will reset everything and fire another initial resize. (This is another place where browsers deviate from the spec - `observer.observe` is supposed to internally call `observer.unobserve` if the element's already being observed, but this didn't work in Firefox, Chrome nor Safari when I actually tried it.)

Reply:
> It turns out Safari is the only browser that actually follows the spec here, though - Firefox and Chrome will always fire an initial resize. Specifically, Safari seems to not fire an initial resize for anything below about 0.018px.

Well that's really bad, because then the associated `WindowEvent`s would just never fire. In addition to the closure not being cleaned up. Right now it's at least not holding any `Rc`s, only `Weak`s, so it shouldn't prevent other things from being cleaned up. I will do some research.

> You could probably also avoid having to create a temporary `ResizeObserver` here by calling `unobserve` then `observe` again on the canvas with the existing `ResizeObserver`, which will reset everything and fire another initial resize.

That's good to know! I will look into it.

Reply:
> It turns out Safari is the only browser that actually follows the spec here, though - Firefox and Chrome will always fire an initial resize. Specifically, Safari seems to not fire an initial resize for anything below about 0.018px.

For reference: https://www.w3.org/TR/resize-observer/#intro
> Observation will fire when observation starts if Element is [being rendered](https://html.spec.whatwg.org/#being-rendered), and Element’s size is not 0,0.

But apparently this has changed in https://github.com/w3c/csswg-drafts/issues/3664 and now observers are required to fire an event no matter the size.
This is also tested by [WPT (test 16)](https://wpt.fyi/results/resize-observer/observe.html?run_id=6263261576560640&run_id=5091417146261504&run_id=5146116692115456&run_id=5116970087743488&run_id=5204317525508096) and works correctly in all browsers except Safari. It's fixed in the new Safari preview though.

I can't come up with anything better then:
1. Use some sort of timeout, which isn't ideal but even worse on Safari because it doesn't support `requestIdleCallback()` either.
2. Detect really small sizes beforehand and don't use oneshot `ResizeObserver`s and instead use `scrollWidth/Height` or `getBoundingClientRect()`.

Leaning more towards proposal 2. I will probably create a tracking issue for all the nonsense we have to work around because old devices can't update their Safari which requires us to continue to support old versions.

Reply:
> I can't come up with anything better then:
> 
>     1. Use some sort of timeout, which isn't ideal but even worse on Safari because it doesn't support `requestIdleCallback()` either.
> 
>     2. Detect really small sizes beforehand and don't use oneshot `ResizeObserver`s and instead use `scrollWidth/Height` or `getBoundingClientRect()`.

I remembered that I completely forgot about `getComputedStyle()` ... I still have to see how consistent it is with the other APIs, but that would be the perfect solution

> > You could probably also avoid having to create a temporary `ResizeObserver` here by calling `unobserve` then `observe` again on the canvas with the existing `ResizeObserver`, which will reset everything and fire another initial resize.
> 
> That's good to know! I will look into it.

I also realized that this could significantly simplify the implementation and I could also split resize and scale change again.

Reply:
Done!

> I also realized that this could significantly simplify the implementation and I could also split resize and scale change again.

Unfortunately this didn't help with that.